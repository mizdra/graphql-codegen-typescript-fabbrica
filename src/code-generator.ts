import type { Config } from './config.js';
import type { ObjectTypeInfo, TypeInfo } from './schema-scanner.js';

function generatePreludeCode(config: Config, typeInfos: TypeInfo[]): string {
  const joinedTypeNames = typeInfos
    .filter(({ type }) => type === 'object')
    .map(({ name }) => `  ${name}`)
    .join(',\n');
  const code = `
import {
  type DefineTypeFactoryInterface${config.nonOptionalDefaultFields ? 'Required' : ''},
  defineTypeFactory,
} from '@mizdra/graphql-codegen-typescript-fabbrica/helper';
import type {
  Maybe,
${joinedTypeNames},
} from '${config.typesFile}';

export * from '@mizdra/graphql-codegen-typescript-fabbrica/helper';
  `.trim();
  return `${code}\n`;
}

export function generateOptionalTypeDefinitionCode(typeInfo: TypeInfo): string {
  if (typeInfo.type === 'object') {
    const { name, fields } = typeInfo;
    const comment = typeInfo.comment ?? '';
    const joinedPropDefinitions = fields
      .map((field) => {
        const comment = field.comment ? `  ${field.comment}` : '';
        return `${comment}  ${field.name}?: ${field.typeString};`;
      })
      .join('\n');
    return `
${comment}export type Optional${name} = {
${joinedPropDefinitions}
};
`.trimStart();
  } else {
    const { name, possibleTypes } = typeInfo;
    const comment = typeInfo.comment ?? '';
    const joinedPossibleTypes = possibleTypes.map((type) => `Optional${type}`).join(' | ');
    return `
${comment}export type Optional${name} = ${joinedPossibleTypes};
`.trimStart();
  }
}

function generateTypeFactoryCode({ nonOptionalDefaultFields }: Config, typeInfo: ObjectTypeInfo): string {
  const { name } = typeInfo;
  return `
/**
 * Define factory for {@link ${name}} model.
 *
 * @param options
 * @returns factory {@link ${name}FactoryInterface}
 */
export const define${name}Factory: DefineTypeFactoryInterface${nonOptionalDefaultFields ? 'Required' : ''}<
  Optional${name},
  {}
> = defineTypeFactory;
`.trimStart();
}

export function generateCode(config: Config, typeInfos: TypeInfo[]): string {
  let code = '';
  code += generatePreludeCode(config, typeInfos);
  code += '\n';
  for (const typeInfo of typeInfos) {
    code += generateOptionalTypeDefinitionCode(typeInfo);
    code += '\n';
    if (typeInfo.type === 'object') {
      code += generateTypeFactoryCode(config, typeInfo);
      code += '\n';
    }
  }
  return code;
}
