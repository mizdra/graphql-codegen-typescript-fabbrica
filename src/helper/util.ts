/** Convert `{ a: number, b: { c: number } }` into `{ a: number | undefined, b: { c: number | undefined } | undefined }`. */
export type DeepOptional<T> = {
  [K in keyof T]: T[K] extends Record<string, unknown> ?
    DeepOptional<T[K]> | undefined // object
  : T[K] extends (infer E)[] ?
    DeepOptional<E>[] | undefined // array
  : T[K] extends readonly (infer E)[] ?
    readonly DeepOptional<E>[] | undefined // readonly array
  : T[K] | undefined; // other
};

export type DeepReadonly<T> =
  T extends Record<string, unknown> | unknown[] | readonly unknown[] ? { readonly [P in keyof T]: DeepReadonly<T[P]> }
  : T;

/** Convert `{ a: number, b: string }` and `{ b: boolean, c: symbol }` into `{ a: number, b: boolean, c: symbol }`. */
export type Merge<F, S> = {
  [K in keyof F | keyof S]: K extends keyof S ?
    S[K] extends Required<S>[K] ? S[K]
    : K extends keyof F ? Required<S>[K] | F[K]
    : S[K]
  : K extends keyof F ? F[K]
  : never;
};
