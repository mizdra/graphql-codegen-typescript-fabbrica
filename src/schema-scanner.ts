import { transformComment } from '@graphql-codegen/visitor-plugin-common';
import type {
  ASTNode,
  GraphQLField,
  GraphQLInputField,
  GraphQLInputObjectType,
  GraphQLObjectType,
  GraphQLSchema,
  GraphQLType,
} from 'graphql';
// NOTE: To avoid `Cannot use GraphQLSchema xxx from another module or realm.`, import from 'graphql/index.js' instead of 'graphql'.
// ref: https://github.com/graphql/graphql-js/issues/1479
import {
  isInputObjectType,
  isInterfaceType,
  isIntrospectionType,
  isListType,
  isNonNullType,
  isObjectType,
  isSpecifiedScalarType,
  isUnionType,
} from 'graphql/index.js';

import type { Config } from './config.js';

// The fork of https://github.com/dotansimha/graphql-code-generator/blob/e1dc75f3c598bf7f83138ca533619716fc73f823/packages/plugins/typescript/resolvers/src/visitor.ts#L85-L91
function clearOptional(str: string): string {
  if (str.startsWith('Maybe')) {
    return str.replace(/Maybe<(.*?)>$/u, '$1');
  }
  return str;
}

// The fork of https://github.com/dotansimha/graphql-code-generator/blob/ba84a3a2758d94dac27fcfbb1bafdf3ed7c32929/packages/plugins/other/visitor-plugin-common/src/base-visitor.ts#L422
function convertName(node: ASTNode | string, config: Config): string {
  let convertedName = '';
  convertedName += config.typesPrefix;
  convertedName += config.convert(node);
  convertedName += config.typesSuffix;
  return convertedName;
}

function isTypeBasedOnUserDefinedType(type: GraphQLType, userDefinedTypeNames: string[]): boolean {
  if (isNonNullType(type)) {
    return isTypeBasedOnUserDefinedType(type.ofType, userDefinedTypeNames);
  } else if (isListType(type)) {
    return isTypeBasedOnUserDefinedType(type.ofType, userDefinedTypeNames);
  } else {
    return userDefinedTypeNames.includes(type.name);
  }
}

function parseTypeNode(type: GraphQLType, config: Config): string {
  if (isNonNullType(type)) {
    return clearOptional(parseTypeNode(type.ofType, config));
  } else if (isListType(type)) {
    return `Maybe<${parseTypeNode(type.ofType, config)}[]>`;
  } else {
    return `Maybe<Optional${convertName(type.name, config)}>`;
  }
}

function parseFieldOrInputField(
  field: GraphQLField<unknown, unknown> | GraphQLInputField,
  convertedTypeName: string,
  config: Config,
  userDefinedTypeNames: string[],
): { typeString: string; comment: string | undefined } {
  const comment = field.description ? transformComment(field.description) : undefined;
  if (isTypeBasedOnUserDefinedType(field.type, userDefinedTypeNames)) {
    return { typeString: `${parseTypeNode(field.type, config)} | undefined`, comment };
  } else {
    return { typeString: `${convertedTypeName}['${field.name}'] | undefined`, comment };
  }
}

function parseObjectTypeOrInputObjectType(
  type: GraphQLObjectType | GraphQLInputObjectType,
  config: Config,
  userDefinedTypeNames: string[],
  getAbstractTypeNames: (type: GraphQLObjectType) => string[],
): ObjectTypeInfo {
  const originalTypeName = type.name;
  const convertedTypeName = convertName(originalTypeName, config);
  const comment = type.description ? transformComment(type.description) : undefined;
  const abstractTypeNames = isObjectType(type) ? getAbstractTypeNames(type) : [];
  const fields = Object.values<GraphQLField<unknown, unknown> | GraphQLInputField>(type.getFields());

  return {
    type: 'object',
    name: convertedTypeName,
    fields: [
      ...(!config.skipTypename ? [{ name: '__typename', typeString: `'${originalTypeName}'` }] : []),
      ...(!config.skipIsAbstractType ?
        abstractTypeNames.map((name) => ({ name: `__is${name}`, typeString: `'${originalTypeName}'` }))
      : []),
      ...fields.map((field) => ({
        name: field.name,
        ...parseFieldOrInputField(field, convertedTypeName, config, userDefinedTypeNames),
      })),
    ],
    comment,
  };
}

type FieldInfo = { name: string; typeString: string; comment?: string | undefined };
export type ObjectTypeInfo = {
  type: 'object';
  name: string;
  fields: FieldInfo[];
  comment?: string | undefined;
};
export type AbstractTypeInfo = {
  type: 'abstract';
  name: string;
  possibleTypes: string[];
  comment?: string | undefined;
};
export type TypeInfo = ObjectTypeInfo | AbstractTypeInfo;

export function getTypeInfos(config: Config, schema: GraphQLSchema): TypeInfo[] {
  const types = Object.values(schema.getTypeMap());

  const userDefinedTypes = types
    .filter((type) => !isIntrospectionType(type) && !isSpecifiedScalarType(type))
    .filter((type) => isObjectType(type) || isInputObjectType(type) || isInterfaceType(type) || isUnionType(type));

  const objectTypes = userDefinedTypes.filter(isObjectType);
  const unionTypes = userDefinedTypes.filter(isUnionType);
  function getAbstractTypeNames(type: GraphQLObjectType): string[] {
    const interfaceNames = type.getInterfaces().map((i) => i.name);
    const unionNames = unionTypes
      .filter((union) => union.getTypes().some((member) => member.name === type.name))
      .map((union) => union.name);
    return [...interfaceNames, ...unionNames];
  }

  const userDefinedTypeNames = userDefinedTypes.map((type) => type.name);

  return userDefinedTypes.map((type) => {
    if (isObjectType(type) || isInputObjectType(type)) {
      return parseObjectTypeOrInputObjectType(type, config, userDefinedTypeNames, getAbstractTypeNames);
    } else if (isInterfaceType(type)) {
      return {
        type: 'abstract',
        name: convertName(type.name, config),
        possibleTypes: objectTypes
          .filter((objectType) => objectType.getInterfaces().some((i) => i.name === type.name))
          .map((objectType) => convertName(objectType.name, config)),
        comment: type.description ? transformComment(type.description) : undefined,
      };
    } else {
      return {
        type: 'abstract',
        name: convertName(type.name, config),
        possibleTypes: type.getTypes().map((member) => convertName(member.name, config)),
        comment: type.description ? transformComment(type.description) : undefined,
      };
    }
  });
}
