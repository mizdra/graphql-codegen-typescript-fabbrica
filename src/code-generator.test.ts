import { transformComment } from '@graphql-codegen/visitor-plugin-common';
import { describe, expect, it } from 'vitest';

import { generateCode, generateOptionalTypeDefinitionCode } from './code-generator.js';
import { TypeInfo } from './schema-scanner.js';
import { fakeConfig, oneOf } from './test/util.js';

describe('generateOptionalTypeDefinitionCode', () => {
  it('generates description comment', () => {
    const typeInfo1: TypeInfo = {
      type: 'object',
      name: 'Book',
      fields: [
        { name: 'id', typeString: 'string | undefined' },
        { name: 'title', typeString: 'string | undefined', comment: transformComment('The book title') },
      ],
      comment: transformComment('The book'),
    };
    expect(generateOptionalTypeDefinitionCode(typeInfo1)).toMatchInlineSnapshot(`
      "/** The book */
      export type OptionalBook = {
        id?: string | undefined;
        /** The book title */
        title?: string | undefined;
      };
      "
    `);
    const typeInfo2: TypeInfo = {
      type: 'abstract',
      name: 'Node',
      possibleTypes: ['Book', 'Author'],
      comment: transformComment('The node'),
    };
    expect(generateOptionalTypeDefinitionCode(typeInfo2)).toMatchInlineSnapshot(`
      "/** The node */
      export type OptionalNode = OptionalBook | OptionalAuthor;
      "
    `);
  });
});

describe('generateCode', () => {
  it('generates code', () => {
    const config = fakeConfig({
      typesFile: './types',
      skipTypename: oneOf([true, false]),
    });
    const typeInfos: TypeInfo[] = [
      {
        type: 'object',
        name: 'Book',
        fields: [
          { name: 'id', typeString: 'string | undefined' },
          { name: 'title', typeString: 'string | undefined' },
          { name: 'author', typeString: 'OptionalAuthor | undefined' },
        ],
      },
      {
        type: 'object',
        name: 'Author',
        fields: [
          { name: 'id', typeString: 'string | undefined' },
          { name: 'name', typeString: 'string | undefined' },
          { name: 'books', typeString: 'Book[] | undefined' },
        ],
      },
      {
        type: 'abstract',
        name: 'Node',
        possibleTypes: ['Book', 'Author'],
      },
    ];
    const actual = generateCode(config, typeInfos);
    expect(actual).toMatchSnapshot();
  });
});
