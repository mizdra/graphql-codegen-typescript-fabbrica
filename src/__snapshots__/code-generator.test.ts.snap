// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`generateCode > generates code 1`] = `
"import {
  type DefineTypeFactoryInterface,
  defineTypeFactory,
} from '@mizdra/graphql-codegen-typescript-fabbrica/helper';
import type {
  Maybe,
  Book,
  Author,
} from './types';

export * from '@mizdra/graphql-codegen-typescript-fabbrica/helper';

export type OptionalBook = {
  id?: string | undefined;
  title?: string | undefined;
  author?: OptionalAuthor | undefined;
};

/**
 * Define factory for {@link Book} model.
 *
 * @param options
 * @returns factory {@link BookFactoryInterface}
 */
export const defineBookFactory: DefineTypeFactoryInterface<
  OptionalBook,
  {}
> = defineTypeFactory;

export type OptionalAuthor = {
  id?: string | undefined;
  name?: string | undefined;
  books?: Book[] | undefined;
};

/**
 * Define factory for {@link Author} model.
 *
 * @param options
 * @returns factory {@link AuthorFactoryInterface}
 */
export const defineAuthorFactory: DefineTypeFactoryInterface<
  OptionalAuthor,
  {}
> = defineTypeFactory;

export type OptionalNode = OptionalBook | OptionalAuthor;

"
`;

exports[`generateCode > generates code when empty typeInfos is passed 1`] = `
"import {
  type DefineTypeFactoryInterface,
  defineTypeFactory,
} from '@mizdra/graphql-codegen-typescript-fabbrica/helper';
import type {
  Maybe,
,
} from './types';

export * from '@mizdra/graphql-codegen-typescript-fabbrica/helper';

"
`;
